crates/primitives/src/signature.rs

```diff
@@ -0,0 +1,855 @@
+    /// Error converting from bytes.
+    FromBytes(&'static str),
+
+    /// Error converting hex to bytes.
+    FromHex(hex::FromHexError),
+
+    /// k256 error
+    #[cfg(feature = "k256")]
+    K256(k256::ecdsa::Error),
+}
+
+#[cfg(feature = "k256")]
+impl From<k256::ecdsa::Error> for SignatureError {
+    fn from(err: k256::ecdsa::Error) -> Self {
+        Self::K256(err)
+    }
+}
+
+impl From<hex::FromHexError> for SignatureError {
+    fn from(err: hex::FromHexError) -> Self {
+        Self::FromHex(err)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for SignatureError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            #[cfg(feature = "k256")]
+            SignatureError::K256(e) => Some(e),
+            SignatureError::FromHex(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl core::fmt::Display for SignatureError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            #[cfg(feature = "k256")]
+            SignatureError::K256(e) => e.fmt(f),
+            SignatureError::FromBytes(e) => f.write_str(e),
+            SignatureError::FromHex(e) => e.fmt(f),
+        }
+    }
+}
+
+/// The parity of the signature, stored as either a V value (which may include
+/// a chain id), or the y-parity.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Parity {
+    /// Explicit V value. May be EIP-155 modified.
+    V(u64),
+    /// Parity flag. True for odd.
+    Parity(bool),
+}
+
+#[cfg(feature = "k256")]
+impl From<k256::ecdsa::RecoveryId> for Parity {
+    fn from(value: k256::ecdsa::RecoveryId) -> Self {
+        Self::Parity(value.is_y_odd())
+    }
+}
+
+impl From<U64> for Parity {
+    fn from(value: U64) -> Self {
+        value.as_limbs()[0].into()
+    }
+}
+
+impl From<Uint<1, 1>> for Parity {
+    fn from(value: Uint<1, 1>) -> Self {
+        Parity::Parity(!value.is_zero())
+    }
+}
+
+impl From<bool> for Parity {
+    fn from(value: bool) -> Self {
+        Parity::Parity(value)
+    }
+}
+
+impl From<u64> for Parity {
+    fn from(value: u64) -> Self {
+        Parity::V(value)
+    }
+}
+
+impl Parity {
+    /// Get the chain_id of the V value, if any
+    pub const fn chain_id(&self) -> Option<ChainId> {
+        match *self {
+            Parity::V(mut v @ 35..) => {
+                if v % 2 == 0 {
+                    v -= 1;
+                }
+                v -= 35;
+                Some(v / 2)
+            }
+            _ => None,
+        }
+    }
+
+    /// Return the y-parity
+    pub fn y_parity(&self) -> bool {
+        match self {
+            Parity::V(v @ 0..=34) => v % 2 == 1,
+            Parity::V(v) => (v ^ 1) % 2 == 0,
+            Parity::Parity(y) => *y,
+        }
+    }
+
+    /// Return the y-parity as 0 or 1
+    pub fn y_parity_byte(&self) -> u8 {
+        self.y_parity() as u8
+    }
+
+    /// Invert the parity
+    pub fn inverted(&self) -> Self {
```

Comment:
fn invert(&mut self)?

Reply:
I've been avoiding mutating functions on signatures, as it becomes too easy to accidentally invalidate them. instead the API takes advantage of `Copy` a lot